File: Leetcode/js/24_swapNodesInPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Your runtime beats 96.55% of javascript submissions. -> 120ms
var swapPairs = function(head) {
    //check input
    if(head === null || head.next === null){
        return head;
    }
    //init
    var prev = new ListNode(0);
    var n1 = head;
    var n2 = head.next;
    var after = head.next.next;
    
    var origHead = n2;
    
    function swap(f,a,b,c) {
        f.next = b;
        a.next = c;
        b.next = a;
    }
    
    //loop
    while(true){
        swap(prev, n1, n2, after);
        //update
        prev = n1;
        n1 = after;
        if(after !== null) {
            n2 = after.next;
        } else {
            break;
        }
        if(n2 !== null) {
            after = n2.next;
        } else {
            break;
        }
    }
    
    return origHead;
};

File: Leetcode/java/24_swapNodesInPairs.java
/*
Given a linked list, swap every two adjacent nodes and return its head.

For example,
Given 1->2->3->4, you should return the list as 2->1->4->3.

Your algorithm should use only constant space. 
You may not modify the values in the list, only nodes itself can be changed.
*/

class Main {
    public static void main(String[] args) {
        System.out.println("Starting...");

        Solution solution = new Solution();

        ListNode input = new ListNode(1);
        input.next = new ListNode(2);
        input.next.next = new ListNode(3);
        input.next.next.next = new ListNode(4);
        
        ListNode result = solution.swapPairs(input);

        System.out.println("Result: " + result.toString());
        
        System.out.println("DONE!");
    }
}


class ListNode {
  int val;
  ListNode next;
  ListNode(int x) {
    val = x;
  }

  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append(this.val);
    ListNode next = this.next;

    while(next != null) {
      sb.append("->").append(next.val);
      next = next.next;
    }

    return sb.toString();
  }
}


class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode preHead = new ListNode(-1);
        ListNode current = preHead;
        preHead.next = head;

        while(current != null) {
            current = swap(current);
        }

        return preHead.next;
    }

    private ListNode swap(ListNode prev){
        // nothing to operate on
        if(prev == null || prev.next == null) {
            return null;
        }

        ListNode first = prev.next;

        // no second item to swap
        if(first.next == null) {
            return null;
        }

        ListNode second = first.next;

        prev.next = second;
        first.next = second.next;
        second.next = first;

        return first;
    }
}












